### ISSUE TITLE
Clarify why multi-stage build work

### ISSUE THREAD

**Issue by Reamer** [openshift/origin#21627]
Not really a problem, just a question. Why does multi-stage build works on openshift 3.10 and possible above? I've searched for a release-note, but I doesn't found anything. I'm happy that multistage builds are working, but I doesn't know if this permanent.
##### Version
```
# oc version
oc v3.10.0+0c4577e-1
kubernetes v1.10.0+b81c8f8
features: Basic-Auth GSSAPI Kerberos SPNEGO
```

```
# docker version
Client:
 Version:         1.13.1
 API version:     1.26
 Package version: docker-1.13.1-75.git8633870.el7.centos.x86_64
 Go version:      go1.9.4
 Git commit:      8633870/1.13.1
 Built:           Fri Sep 28 19:45:08 2018
 OS/Arch:         linux/amd64

Server:
 Version:         1.13.1
 API version:     1.26 (minimum version 1.12)
 Package version: docker-1.13.1-75.git8633870.el7.centos.x86_64
 Go version:      go1.9.4
 Git commit:      8633870/1.13.1
 Built:           Fri Sep 28 19:45:08 2018
 OS/Arch:         linux/amd64
 Experimental:    false
```

##### Steps To Reproduce
1. Install Openshift
2. Create a BuildConfig with multistage instructions and ImageStream for pushing to registry (Example: https://gist.github.com/Reamer/d1393d42249b328fc17e0cb6863a38f5)

##### Current Result

It works.

##### Expected Result

Related with https://github.com/openshift/origin/issues/19198. It should not work.

##### Additional Information
Building the same Dockerfile directly on node with "docker build ..." doesn't work.

**Comment by merusso** [comment]
I'm also mystified by this. Can anyone explain this?

**Comment by ofrzeta** [comment]
I tried it on OpenShift Enterprise 3.11 and it also worked. Here's a Gist with busybox and appropriate Entrypoint added, so you can rsh to the pods https://gist.github.com/ofrzeta/3bb08efc79f4c33fa267835b72bc96f6

**Comment by ofrzeta** [comment]
The build is run in the build container with a specific builder image (https://github.com/openshift/builder). Therefore it is not using the underlying Docker. When you debug the build pod (oc debug mytestbuild-1-build) you can run the command "openshift-docker-build --log-level=5" you can see that the tool detects the multi-stage configuration:
```
Starting Docker build from build config mytestbuild-1 ...
Detected multi-stage Dockerfile, image will be built with imageOptimizationPolicy set to SkipLayers
```

When you check the build pod configuration you can see the corresponding image registry URL:

`image: registry.redhat.io/openshift3/ose-docker-builder:v3.11.59`

**Comment by sbose78** [comment]
@bparees , Does the openshift ci infra support multi-stage builds?

```
images: 
- from: os
  to: devconsole-operator
  dockerfile_path: Dockerfile
```
  
( https://github.com/redhat-developer/devconsole-operator/blob/master/Dockerfile )


caused the following error ..

```
Pulling image build-tools ...
error: build error: failed to pull image: repository docker.io/build-tools not found: does not exist or no pull access
2019/04/02 07:04:18 Ran for 4m56s
error: could not run steps: could not wait for build: the build devconsole-operator failed after 8s with reason PullBuilderImageFailed: Failed pulling builder image.

Pulling image build-tools ...
error: build error: failed to pull image: repository docker.io/build-tools not found: does not exist or no pull access
```

Wondering if I'm not writing my Dockerfile right?

**Comment by bparees** [comment]
@sbose78 yes it supports multistage builds.    what is "docker.io/build-tools"?
```
docker pull docker.io/build-tools
Using default tag: latest
Error response from daemon: pull access denied for build-tools, repository does not exist or may require 'docker login'
```

Seems like maybe you expected that value to be replaced with something?

**Comment by sbose78** [comment]
Thank you!

Here's what is causing that. My Dockerfile looks like:
```
FROM centos:7 as build-tools
....
...
..
FROM build-tools as builder
...
....
..
```

@bparees  That's coming from https://github.com/redhat-developer/devconsole-operator/blob/master/Dockerfile

**Comment by bparees** [comment]
@sbose78 can you point me to your ci-operator config?  Do you have someone from our test platform team helping you get this setup?

**Comment by bparees** [comment]
(we should also take this discussion elsewhere, this issue isn't the right spot for it.  aos-devel is a good mailing list for it)

**Comment by bguruprasad** [comment]
> @bguruprasad can you open this as a bugzilla at bugzilla.redhat.com? That is a better way for us to track it rather than using a closed issue. Also please include what version of openshift you are using.
> 
> might be useful to see the error you're getting when you try to run the image, and have you done any inspection of the jar file from the working image vs the non-working image? different sizes? different permissions? etc.

Sure @bparees  I'll do that. 

Yes, I did the initial investigation, everything looks good but still the image fails to run with error below. I'll put the detailed bug in bugzilla as suggested. Thanks.

![image](https://user-images.githubusercontent.com/3767545/65526454-08ece580-dee9-11e9-95f9-ef9548304924.png)

**Comment by Piroddi** [comment]
Hi @bguruprasad, could you please share your bug number. I am running into the same error. Thank you

**Comment by hfuss** [comment]
@sbose78 did you ever find a solution? I'm having the same issue

```
failed to pull image: repository docker.io/install not found: does not exist or no pull access
```

where `install` is the name of one of my stages in my docker build:

```dockerfile
FROM openjdk:8u181-jre-slim as install
# ...
FROM install
```

OpenShift version:
```
openshift v3.11.98
kubernetes v1.11.0+d4cacc0
```

docker API version: 1.13.1

**Comment by dAnjou** [comment]
Hi @sbose78 @bparees, did you continue investigating somewhere else? I'm also running into this issue now :(

**Comment by bparees** [comment]
@dAnjou i don't believe a bug ever got opened for us to investigate this.  if you are interested in helping us pursue it, please do so (bugzilla.redhat.com, OCP product, build component) and link this thread.

But just to temper expectations, 3.11 is in maintenance mode, so if the issue can't be recreated in 4.x (and it likely can't, since the image build engine was entirely replaced in 4.x), there's likely going to be limited appetite to work on a fix, vs trying to give you a workaround.

fundamentally this is likely a bug in imagebuilder(what 3.11 uses under the covers to build images when you have a multistage dockerfile), so you might see if you can recreate it w/ that, which would help a lot:
https://github.com/openshift/imagebuilder/

in particular you might try the level of imagebuilder that 3.11 uses:
https://github.com/openshift/origin/blob/bf985b15a81c254c52dc86d752367e2b311e36c0/glide.lock#L977-L978

if you can demonstrate this is as simple as updating 3.11 to the latest imagebuilder, that'd make it much more likely to get fixed.

cc @adambkaplan

**Comment by data-ux** [comment]
A workaround that fixed the issue for me was using an explicit path in the COPY command (instead of wildcard):
`COPY --from=builder /usr/app/build/libs/actual-name-of-file.jar application.jar`

See discussion here: https://bugzilla.redhat.com/show_bug.cgi?id=1798079#c4
